Add <QLoader> component

purepennons opened the pull request:
Resolved #1 
- - -
# Done
- `<Loader>`: component for logic
- `<QLoader>`: component for guideline loader
- `<QSpinner>`: guideline spinner

# Todo
- [ ] timeout modal
- [ ] document 
- [ ] test case
- [ ] fix loader circle animation broken

# Preview
<img width="785" alt="screen shot 2017-12-16 at 2 34 58 am" src="https://user-images.githubusercontent.com/4588113/34055586-d57a66be-e20a-11e7-9c19-8d5e8ba0c9f8.png">

<img width="789" alt="screen shot 2017-12-16 at 2 35 16 am" src="https://user-images.githubusercontent.com/4588113/34055585-d54eef16-e20a-11e7-8b87-8f11c96be752.png">

purepennons committed:
loader: initial commit

purepennons committed:
add normalize.css, ramda and rxjs

purepennons committed:
feat: Loader component is work (without styles)

purepennons committed:
feat: Loader component supports timeout prop

purepennons committed:
refactor: remove built files

purepennons committed:
fix: remove color-hex-length rule from stylelint config

purepennons committed:
feat: pass props and state to render props

purepennons committed:
feat: add percentage and icon spinner

purepennons committed:
feat: icon spinner adds png version

purepennons committed:
feat: percent prop of Spinner accepts both number and string types

purepennons committed:
feat: remove unused class name to prevent css naming conflict

purepennons committed:
add react-qui-button and update akiya-react-component-scripts to 3.2.0

purepennons committed:
feat: QLoader basically works

#1

purepennons committed:
refactor: use rxjs only to prevent memory issue

purepennons committed:
refactor: remove ramda from Loader

#1

purepennons committed:
chore: add lib and build/ to .gitignore

purepennons committed:
feat: also export Spinner component as QSpinner

purepennons committed:
feat: pass rest props to <Spinner>, <Loader> and <QLoader>

Chun-Lin commented on packages/loader/src/react-component-node-modules/components/Loader/Loader.js:
```diff
@@ -0,0 +1,129 @@
+    this.resetState()
+    this.timerSub = this.getPhasesObservableSource(
+      this.state.renderQueue,
+    ).subscribe(render => this.setState(_ => ({ currentRender: render })))
```
`_ => ({currentRender: render})` 跟 `() => ({currentRender: render})`的用法是一樣的嗎？

Chun-Lin commented on packages/loader/src/react-component-node-modules/components/Loader/Loader.js:
```diff
@@ -0,0 +1,129 @@
+  enableLoader() {
+    this.resetState()
+    this.timerSub = this.getPhasesObservableSource(
+      this.state.renderQueue,
```
`this.state.renderQueue,`可以拿掉

purepennons replied:
fixed

Chun-Lin commented on packages/loader/src/react-component-node-modules/components/Loader/Loader.js:
```diff
@@ -0,0 +1,129 @@
+    this.resetState()
+    this.timerSub = this.getPhasesObservableSource(
+      this.state.renderQueue,
+    ).subscribe(render => this.setState(_ => ({ currentRender: render })))
```
`_ => ({currentRender: render})` 跟 `() => ({currentRender: render})`的用法是一樣的嗎？

purepennons replied:
setState 有吃物件與 callback function 兩種用法。弔詭的是，不管是哪種寫法，setState 在 react 底層都是非同步的行為（anti-pattern）。因此如果你的 setState 要更新的資料是相依之前 state 的數值，會導致 update 的值不一定正確。 

而 callback 形式的，則可以確保拿到上一次 state 的物件是正確的，並且符合非同步的書寫習慣。已經有討論未來可能會棄用 Object 形式的 setState。所以還是從現在養成正確的書寫習慣吧，即便不需要上一次的 state，不用到的變數，我們習慣上是以 `_` 表達。 

[https://medium.com/javascript-scene/setstate-gate-abc10a9b2d82](https://medium.com/javascript-scene/setstate-gate-abc10a9b2d82)

purepennons committed:
fix: use invariant to handle error and remove rest props to prevent warnings

  - add invariant and @react-qui/button packages
  - handle error by invariant pacakge
  - remove rest props in <Loader> and <QLoader>
